### PR TITLE
Refactor interaction code and add stdin support

### LINKED ISSUE #6: Handle stdin

### PR DESCRIPTION
**Pull request by zunda-arrow**
Refactors interaction code:
	- Rename uses of "button" to "component" when appropriate. 
	- Use enums and cases to make it more clear what IDs are and if statements are exhaustive.
	- Change the selectors tuple to a `Selector` named tuple with documentation to improve clarity.

Stdin support is self explanatory. 

closes #6

### CODE REVIEW

File: bot/plugins/instance.py [hunk 1]
```diff
@@ -105,58 +130,74 @@ async def on_button(event: hikari.InteractionCreateEvent) -> None:
...
+        case ComponentID.VERSION:
+            if v := event.interaction.values:
+                inst.version = v[0]
+            else:
+                inst.version = None
+        case ComponentID.STDIN:
+            await event.app.rest.create_modal_response(
+                event.interaction,
+                event.interaction.token,
+                title="Set Stdin",
```
Comment (circuitsacul) [inline bot/plugins/instance.py]: Set STDIN?

File: bot/plugins/instance.py [hunk 2]
```diff
@@ -105,58 +130,74 @@ async def on_button(event: hikari.InteractionCreateEvent) -> None:
...
+                custom_id=ModalID.STDIN,
+                component=event.app.rest.build_modal_action_row().add_text_input(
+                    ModalID.STDIN,
+                    "Set Stdin",
```
Comment (circuitsacul) [inline bot/plugins/instance.py]: Set STDIN?

File: bot/plugins/instance.py [hunk 3]
```diff
@@ -342,29 +409,36 @@ def components(self) -> list[hikari.api.MessageActionRowBuilder]:
...
             .add_interactive_button(
                 hikari.ButtonStyle.SECONDARY,
-                "language",
+                ComponentID.LANGUAGE,
                 label=f"Language: {self.language.v or 'Unknown'}",
             )
+            .add_interactive_button(
+                hikari.ButtonStyle.SECONDARY,
+                ComponentID.STDIN,
+                label="stdin",
```
Comment (circuitsacul) [inline bot/plugins/instance.py]: STDIN?

**Review by circuitsacul — APPROVED** [review]
Also it would probably be a good idea to show the currently used STDIN alongside the codeblock (as another code block below it). The idea being, other people can see what you used.

If this happens, some code changes are needed in models.py#475-480. The STDIN and code should be seperate if they're uploaded as files, and I think stdin should be checked first (so prefer putting stdin in a file and keeping the code as a code block)

File: bot/plugins/instance.py [hunk 4]
```diff
@@ -395,41 +469,69 @@ async def execute(self) -> None:
...
-            if len(out) > 1_950:
-                att = hikari.Bytes(out, "output.ansi")
-                out = None
-            else:
-                if out in {"", "\n"}:
-                    out = "Your code ran with no output."
-                else:
-                    out = f"```ansi\n{out}\n```"
+            if not code_output.strip():
+                out += ["Your code ran with no output."]
```
Comment (circuitsacul) [inline bot/plugins/instance.py]:
```suggestion
                out.append("Your code ran with no output.")
```

File: bot/plugins/instance.py [hunk 5]
```diff
@@ -395,41 +469,69 @@ async def execute(self) -> None:
...
+                code_output = ""
         else:
-            out = "No runtime selected."
+            out += ["No runtime selected."]
```
Comment (circuitsacul) [inline bot/plugins/instance.py]:
```suggestion
            out.append("No runtime selected.")
```

File: bot/plugins/instance.py [hunk 6]
```diff
@@ -395,41 +469,69 @@ async def execute(self) -> None:
...
 
-        if out:
-            out = f"<@{self.requester}>\n{out}"
-        else:
-            out = f"<@{self.requester}>"
+        stdin: str = self.stdin or ""
```
Comment (circuitsacul) [inline bot/plugins/instance.py]:
```suggestion
        stdin = self.stdin or ""
```

File: bot/plugins/instance.py [hunk 7]
```diff
@@ -395,41 +469,69 @@ async def execute(self) -> None:
...
+            else:
+                out += [f"```ansi\n{code_output}```"]
+
+        if stdin:
+            if stdin_in_file:
+                stdin_attr = hikari.Bytes(stdin, "stdin.txt")
+            else:
+                stdin = "\n".join(
+                    f"\x1b[1;33mIN:\x1b[0m {line}" for line in stdin.splitlines()
+                )
```
Comment (circuitsacul) [inline bot/plugins/instance.py]: This at the very least needs to happen before the len() check. An input could be 1800 newlines, which would be under 1950, but far over with all the formatting this adds.

I'm not sure if it's a good idea to also have this formatting in the attachments, but it's worth noting that if the file extension is .ansi it will support ansi the same way codeblocks do.

File: bot/plugins/instance.py [hunk 8]
```diff
@@ -395,41 +469,69 @@ async def execute(self) -> None:
...
+                code_output_in_file = True
+            else:
+                stdin_in_file = True
+                code_output_in_file = True
+
+        if code_output:
+            if code_output_in_file:
+                code_attr = hikari.Bytes(code_output, "code.ansi")
+            else:
+                out += [f"```ansi\n{code_output}```"]
```
Comment (circuitsacul) [inline bot/plugins/instance.py]:
```suggestion
                out.append(f"```ansi\n{code_output}```")
```

File: bot/plugins/instance.py [hunk 9]
```diff
@@ -395,41 +469,69 @@ async def execute(self) -> None:
             await plugin.app.rest.trigger_typing(self.channel)
 
         # try to execute code
-        att: hikari.Bytes | None = None
-        out: str | None
+        code_attr: hikari.Bytes | None = None
+        stdin_attr: hikari.Bytes | None = None
```
Comment (circuitsacul) [inline bot/plugins/instance.py]: This was probably done out of habit but these should be `att` or `attachment`, rather than `attr` (which stands for "attribute").

A better name would probably be `code_file` and `stdin_file` though.

File: bot/plugins/instance.py [hunk 10]
```diff
@@ -395,41 +469,69 @@ async def execute(self) -> None:
...
+        out: list[str] = [f"<@{self.requester}>"]
+
+        code_output_in_file = False
+        stdin_in_file = False
```
Comment (circuitsacul) [inline bot/plugins/instance.py:477]: Isn't this redundant, since code_file and stdin_file are None for false and Bytes for true?

**Review by circuitsacul — CHANGES_REQUESTED** [review]

File: bot/plugins/instance.py (hunk 10 above)
Comment (zunda-arrow) [reply bot/plugins/instance.py:477]: Doing it like this means I only need to assign the bytes class to a variable in one place.

Comment (circuitsacul) [reply bot/plugins/instance.py:477]: Oh I see

File: bot/plugins/instance.py (hunk 9 above)
Comment (zunda-arrow) [reply bot/plugins/instance.py]: changed

File: bot/plugins/instance.py (hunk 7 above)
Comment (zunda-arrow) [reply bot/plugins/instance.py]: I don't think its a good idea to put the formatting in the file. Its in the message to add clarity, I don't think it adds clarity when the file is already named `stdin.txt`.

Comment (circuitsacul) [reply bot/plugins/instance.py]: Fair enough. Still need to handle the len() thing though

File: bot/plugins/instance.py [hunk 11]
```diff
@@ -395,41 +469,69 @@ async def execute(self) -> None:
...
+            else:
+                out.append(f"```ansi\n{code_output}```")
+
+        if stdin:
+            if stdin_in_file:
+                stdin_file = hikari.Bytes(stdin, "stdin.txt")
+            else:
+                stdin = "\n".join(
+                    f"\x1b[1;33mIN:\x1b[0m {line}" for line in stdin.splitlines()
+                )
```
Comment (circuitsacul) [inline bot/plugins/instance.py]: untested solution

```suggestion
        stdin = self.stdin or ""
        stdin_formatted = "\n".join(
            f"\x1b[1;33mIN:\x1b[0m {line}" for line in stdin.splitlines()
        )

        if len(code_output) + len(stdin_formatted) > 1_950:
            if len(code_output) < 1_950:
                stdin_in_file = True
            elif len(stdin_formatted) < 1_950:
                code_output_in_file = True
            else:
                code_output_in_file = True
                stdin_in_file = True

        if code_output:
            if code_output_in_file:
                code_file = hikari.Bytes(code_output, "code.ansi")
            else:
                out.append(f"```ansi\n{code_output}```")

        if stdin:
            if stdin_in_file:
                stdin_file = hikari.Bytes(stdin, "stdin.txt")
            else:
                stdin = stdin_formatted
```

Comment (zunda-arrow) [reply bot/plugins/instance.py]: wrote my own code that i tested so im gonna resolve this

File: bot/plugins/instance.py [hunk 12]
```diff
@@ -395,41 +469,69 @@ async def execute(self) -> None:
...
+            if code_output_in_file:
+                code_file = hikari.Bytes(code_output, "code.ansi")
+            else:
+                out.append(f"```ansi\n{code_output}```")
+
+        if stdin:
+            if stdin_in_file:
+                stdin_file = hikari.Bytes(stdin, "stdin.txt")
+            else:
+                out += [f"```ansi\n{formatted_stdin}```"]
```
Comment (circuitsacul) [inline bot/plugins/instance.py]:
```suggestion
                out.append(f"```ansi\n{formatted_stdin}```")
```

File: bot/app.py [hunk 13]
```diff
@@ -20,7 +20,7 @@ async def shutdown(self) -> None:
 
 
 def main() -> None:
-    app = hikari.GatewayBot(CONFIG.TOKEN, intents=INTENTS)
+    app = hikari.GatewayBot(CONFIG.TOKEN, intents=INTENTS, banner="crescent")
```
Comment (circuitsacul) [inline bot/app.py]: This is out of scope for this PR. Personally I prefer to keep the hikari banner, but I'm willing to discuss is in an issue and possibly change it in a separate pr

```suggestion
    app = hikari.GatewayBot(CONFIG.TOKEN, intents=INTENTS)
```

**Review by circuitsacul — CHANGES_REQUESTED** [review]

File: bot/plugins/instance.py (hunk 12 above)
Comment (circuitsacul) [reply bot/plugins/instance.py]: since you merged the other suggestions I went ahead and just added this

File: bot/app.py (hunk 13 above)
Comment (zunda-arrow) [reply bot/app.py]: ohhh fuck i didn't think i saved that lmao

**Review by circuitsacul — APPROVED** [review]
tysm for this
